File: llm/providers/fallback.ts
import type { ModelCandidate } from "./provider-registry.js";

export type FailoverReason =
  | "auth" | "rate_limit" | "overloaded" | "server_error"
  | "model_not_found" | "provider_error" | "format"
  | "connection_reset" | "timeout" | "empty_stream" | "unknown";

export type ErrorPermanence = "transient" | "permanent";

export interface ErrorClassification {
  reason: FailoverReason;
  permanence: ErrorPermanence;
  statusCode?: number;
  message?: string;
}

const JITTER_MIN_MS = 1000;
const JITTER_MAX_MS = 3000;
const DEFAULT_TIMEOUT_MS = 120_000;

export function classifyError(error: unknown, providerErrorCodes: string[] = []): ErrorClassification {
  const err = error as any;
  const statusCode: number | undefined = err?.status ?? err?.statusCode ?? err?.error?.status;
  const message: string = err?.message ?? String(error);
  const errorCode: string | undefined = err?.error?.error?.type ?? err?.error?.error?.code;

  if (err?.code === "ECONNRESET" || err?.code === "ETIMEDOUT" || err?.code === "ECONNREFUSED" || err?.code === "EPIPE") {
    return { reason: "connection_reset", permanence: "transient", message };
  }
  if (message.includes("timeout") || err?.code === "UND_ERR_CONNECT_TIMEOUT") {
    return { reason: "timeout", permanence: "transient", message };
  }
  if (statusCode) {
    if (statusCode === 401 || statusCode === 403) return { reason: "auth", permanence: "permanent", statusCode, message };
    if (statusCode === 429) return { reason: "rate_limit", permanence: "transient", statusCode, message };
    if (statusCode === 529) return { reason: "overloaded", permanence: "transient", statusCode, message };
    if (statusCode >= 500) return { reason: "server_error", permanence: "transient", statusCode, message };
    if (statusCode === 400) {
      if (errorCode && providerErrorCodes.length > 0 && providerErrorCodes.includes(errorCode))
        return { reason: "provider_error", permanence: "transient", statusCode, message };
      if (errorCode === "invalid_request_error" || !errorCode)
        return { reason: "format", permanence: "permanent", statusCode, message };
      return { reason: "provider_error", permanence: "transient", statusCode, message };
    }
    if (statusCode === 404) return { reason: "model_not_found", permanence: "permanent", statusCode, message };
  }
  return { reason: "unknown", permanence: "transient", message };
}

export interface FallbackAttempt {
  providerId: string;
  model: string;
  reason: FailoverReason | null;
  permanence?: ErrorPermanence;
  durationMs: number;
  retriesUsed: number;
  timestamp: string;
}

export interface FallbackResult<T> {
  data: T;
  attempts: FallbackAttempt[];
  usedCandidate: ModelCandidate;
}

export class FallbackExhaustedError extends Error {
  attempts: FallbackAttempt[];
  constructor(attempts: FallbackAttempt[]) {
    const summary = attempts.map((a) => `  ${a.providerId}/${a.model}: ${a.reason}`).join("\n");
    super(`所有 provider 均失败:\n${summary}`);
    this.name = "FallbackExhaustedError";
    this.attempts = attempts;
  }
}

function sleepWithJitter(minMs: number, maxMs: number): Promise<void> {
  const delay = minMs + Math.random() * (maxMs - minMs);
  return new Promise((resolve) => setTimeout(resolve, delay));
}

function withTimeout<T>(promise: Promise<T>, ms: number): Promise<T> {
  return new Promise<T>((resolve, reject) => {
    const timer = setTimeout(() => reject(new Error("timeout")), ms);
    promise.then(
      (val) => { clearTimeout(timer); resolve(val); },
      (err) => { clearTimeout(timer); reject(err); },
    );
  });
}

export async function tryProviderWithFallback<T>(
  candidates: ModelCandidate[],
  action: (candidate: ModelCandidate) => Promise<T>,
  onAttempt?: (attempt: FallbackAttempt) => void,
): Promise<FallbackResult<T>> {
  const attempts: FallbackAttempt[] = [];

  for (const candidate of candidates) {
    const start = Date.now();
    try {
      const result = await withTimeout(action(candidate), DEFAULT_TIMEOUT_MS);
      const attempt: FallbackAttempt = {
        providerId: candidate.providerId, model: candidate.model, reason: null,
        durationMs: Date.now() - start, retriesUsed: 0, timestamp: new Date().toISOString(),
      };
      attempts.push(attempt);
      onAttempt?.(attempt);
      return { data: result, attempts, usedCandidate: candidate };
    } catch (error) {
      const classification = classifyError(error);
      const durationMs = Date.now() - start;

      if (classification.reason === "rate_limit") {
        try {
          await sleepWithJitter(JITTER_MIN_MS, JITTER_MAX_MS);
          const result = await withTimeout(action(candidate), DEFAULT_TIMEOUT_MS);
          const attempt: FallbackAttempt = {
            providerId: candidate.providerId, model: candidate.model, reason: null,
            durationMs: Date.now() - start, retriesUsed: 1, timestamp: new Date().toISOString(),
          };
          attempts.push(attempt);
          onAttempt?.(attempt);
          return { data: result, attempts, usedCandidate: candidate };
        } catch { /* 重试也失败，继续 fallback */ }
      }

      const attempt: FallbackAttempt = {
        providerId: candidate.providerId, model: candidate.model, reason: classification.reason,
        permanence: classification.permanence, durationMs, retriesUsed: 0, timestamp: new Date().toISOString(),
      };
      attempts.push(attempt);
      onAttempt?.(attempt);

      if (classification.permanence === "permanent") break;
    }
  }

  throw new FallbackExhaustedError(attempts);
}

File: src/main/scala/nebscala/llm/fallback.scala
package nebscala.llm

import cats.effect.{IO, Temporal}
import cats.syntax.all.*
import scala.concurrent.duration.*
import scala.util.Random

enum FailoverReason:
  case Auth, RateLimit, Overloaded, ServerError, ModelNotFound, ProviderError, Format, ConnectionReset, Timeout, EmptyStream, Unknown

enum ErrorPermanence:
  case Transient, Permanent

case class ErrorClassification(
  reason: FailoverReason,
  permanence: ErrorPermanence,
  statusCode: Option[Int] = None,
  message: Option[String] = None
)

case class FallbackAttempt(
  providerId: String,
  model: String,
  reason: Option[FailoverReason],
  permanence: Option[ErrorPermanence],
  durationMs: Long,
  retriesUsed: Int,
  timestamp: String
)

case class FallbackResult[T](
  data: T,
  attempts: List[FallbackAttempt],
  usedCandidate: ModelCandidate
)

class FallbackExhaustedError(val attempts: List[FallbackAttempt]) extends Exception:
  override def getMessage: String =
    val summary = attempts.map { a =>
      s"  ${a.providerId}/${a.model}: ${a.reason.map(_.toString).getOrElse("unknown")}"
    }.mkString("\n")
    s"All providers failed:\n$summary"

object Fallback:
  private val JitterMinMs = 1000
  private val JitterMaxMs = 3000
  private val DefaultTimeoutMs = 120_000

  def classifyError(error: Throwable): ErrorClassification =
    val msg = error.getMessage.toLowerCase

    if msg.contains("connection reset") || msg.contains("econnreset") || msg.contains("econnrefused") || msg.contains("epipe") || msg.contains("broken pipe") then
      ErrorClassification(FailoverReason.ConnectionReset, ErrorPermanence.Transient, message = Some(error.getMessage))
    else if msg.contains("timeout") || msg.contains("timed out") then
      ErrorClassification(FailoverReason.Timeout, ErrorPermanence.Transient, message = Some(error.getMessage))
    else if msg.contains("auth") || msg.contains("unauthorized") || msg.contains("403") || msg.contains("401") then
      ErrorClassification(FailoverReason.Auth, ErrorPermanence.Permanent, message = Some(error.getMessage))
    else if msg.contains("rate limit") || msg.contains("429") then
      ErrorClassification(FailoverReason.RateLimit, ErrorPermanence.Transient, message = Some(error.getMessage))
    else if msg.contains("overloaded") || msg.contains("529") then
      ErrorClassification(FailoverReason.Overloaded, ErrorPermanence.Transient, message = Some(error.getMessage))
    else if msg.contains("server error") || msg.contains("500") || msg.contains("502") || msg.contains("503") then
      ErrorClassification(FailoverReason.ServerError, ErrorPermanence.Transient, message = Some(error.getMessage))
    else if msg.contains("model not found") || msg.contains("404") then
      ErrorClassification(FailoverReason.ModelNotFound, ErrorPermanence.Permanent, message = Some(error.getMessage))
    else if msg.contains("invalid request") || msg.contains("bad request") || msg.contains("400") then
      ErrorClassification(FailoverReason.Format, ErrorPermanence.Permanent, message = Some(error.getMessage))
    else
      ErrorClassification(FailoverReason.Unknown, ErrorPermanence.Transient, message = Some(error.getMessage))

  private def withTimeoutIO[A](ioa: IO[A], ms: Long): IO[A] =
    ioa.timeout(ms.millis).handleErrorWith { err =>
      IO.raiseError(new Exception(s"timeout", err))
    }

  def sleepWithJitter(minMs: Int, maxMs: Int): IO[Unit] =
    val delay = minMs + Random.nextInt(maxMs - minMs)
    IO.sleep(delay.millis)

  def tryProviderWithFallback[T](
    candidates: List[ModelCandidate],
    action: ModelCandidate => IO[T],
    onAttempt: Option[FallbackAttempt => IO[Unit]] = None
  ): IO[FallbackResult[T]] =
    def loop(
      remaining: List[ModelCandidate],
      attempts: List[FallbackAttempt]
    ): IO[FallbackResult[T]] = remaining match
      case Nil =>
        IO.raiseError(new FallbackExhaustedError(attempts))
      case candidate :: rest =>
        val start = System.currentTimeMillis()
        val attemptIO = withTimeoutIO(action(candidate), DefaultTimeoutMs)

        attemptIO.attempt.flatMap {
          case Right(result) =>
            val attempt = FallbackAttempt(
              candidate.providerId, candidate.model, None, None,
              System.currentTimeMillis() - start, 0,
              java.time.Instant.now().toString
            )
            val newAttempts = attempts :+ attempt
            onAttempt.traverse_(_.apply(attempt)) *>
              IO.pure(FallbackResult(result, newAttempts, candidate))
          case Left(error) =>
            val classification = classifyError(error)
            val durationMs = System.currentTimeMillis() - start

            // Rate limit: retry with jitter
            val withRetry: IO[FallbackResult[T]] =
              if classification.reason == FailoverReason.RateLimit then
                sleepWithJitter(JitterMinMs, JitterMaxMs) *>
                  withTimeoutIO(action(candidate), DefaultTimeoutMs).attempt.flatMap {
                    case Right(result) =>
                      val attempt = FallbackAttempt(
                        candidate.providerId, candidate.model, None, None,
                        System.currentTimeMillis() - start, 1,
                        java.time.Instant.now().toString
                      )
                      val newAttempts = attempts :+ attempt
                      onAttempt.traverse_(_.apply(attempt)) *>
                        IO.pure(FallbackResult(result, newAttempts, candidate))
                    case Left(_) =>
                      val failAttempt = FallbackAttempt(
                        candidate.providerId, candidate.model,
                        Some(classification.reason), Some(classification.permanence),
                        durationMs, 0, java.time.Instant.now().toString
                      )
                      val newAttempts = attempts :+ failAttempt
                      onAttempt.traverse_(_.apply(failAttempt)) *>
                        (if classification.permanence == ErrorPermanence.Permanent then
                          IO.raiseError(new FallbackExhaustedError(newAttempts))
                        else
                          loop(rest, newAttempts))
                  }
              else
                val attempt = FallbackAttempt(
                  candidate.providerId, candidate.model,
                  Some(classification.reason), Some(classification.permanence),
                  durationMs, 0, java.time.Instant.now().toString
                )
                val newAttempts = attempts :+ attempt
                onAttempt.traverse_(_.apply(attempt)) *>
                  (if classification.permanence == ErrorPermanence.Permanent then
                    IO.raiseError(new FallbackExhaustedError(newAttempts))
                  else
                    loop(rest, newAttempts))

            withRetry
        }

    loop(candidates, Nil)
